File: bit&seg/Dquery.cs
}
using System;
using System.Linq;
using System.Collection.Generics;

class Dquery {
	static long[] bt;
    static readonly char[] split = { ' ' };
    private const int MAX = 10000001;

    public static void Main(string[] args)
    {
        var occ = new int[MAX];
        for (var i = 0; i < MAX; i++)
        {
            occ[i] = -1;
        }

        int n = readInt();
        var num = readArr();
        bt = new long[n + 1];
        int qc = readInt();
        var q = new QData[qc];
        for (int i = 0; i < qc; i++)
        {
            var ip = readArr();
            q[i] = new QData(ip[0], ip[1], i);
        }
        Comparison<QData> cmp = (q1, q2) => q1.j.CompareTo(q2.j);
        Array.Sort(q, cmp);
        var result = new long[qc];
        var pos = 0;
        for (var i = 0; i < qc; i++)
        {
            while (pos < n && pos <= q[i].j)
            {
                if (occ[num[pos]] != -1)
                {
                    update(occ[num[pos]] + 1, -1);
                }
                update(pos + 1, 1);
                occ[num[pos]] = pos++;
            }

            result[q[i].pos] = query(q[i].j) - query(q[i].i - 1);
        }

        for (var i = 0; i < result.Length; i++)
        {
            Console.WriteLine(result[i]);
        }
    }

    class QData
    {
        public int i;
        public int j;
        public int pos;
        public QData(int i, int j, int pos)
        {
            this.i = i;
            this.j = j;
            this.pos = pos;
        }
    }
    static int readInt()
    {
        return int.Parse(Console.ReadLine().Trim());
    }

    static int[] readArr()
    {
        return Console.ReadLine().Trim().Split(split).Select(n => int.Parse(n)).ToArray();
        /*var s = Console.ReadLine().Trim().Split(new char[] {' '});
        var r = new Data[s.Length];
        for(int i = 0; i < s.Length; i++) {
            r[i] = new Data(i, int.Parse(s[i]));
        }
        return r;*/
    }

    static void update(int i, int v)
    {
        while (i < bt.Length)
        {
            bt[i] += v;
            i += (i & -i);
        }
    }

    static long query(int i)
    {
        long r = 0;
        while (i > 0)
        {
            r += bt[i];
            i -= (i & -i);
        }
        return r;
    }

File: bit&seg/Dquery.cpp
#include <iostream>
#include <cstdlib>
#include <algorithm>
using namespace std;
int bt[30010];
#define MAXN 1000010
#define MAXVAL 30001

void update(int i, int k){
    while(i<=MAXVAL){
        bt[i]+=k;
        i += (i & -i);
    }
}

int query(int i){
    int s = 0;
    while(i){
        s += bt[i];
        i -= (i & -i);
    }
    return s;
}

int range(int i, int j) {
	return query(j) - query(i-1);
}


struct qdata{
    long i , j , indx;
};


bool c1(const qdata &a , const qdata &b){
    return a.j < b.j;
}

int main(){
    int n;
    scanf("%d",&n);
    int d[n];
    int occ[MAXN];

    for(int i = 0; i < MAXN; i++) {
        occ[i] = -1;
    }

    for(int j = 0 ; j < n ; j++) {
        scanf("%d",&d[j]);
    }

    int qn;
    scanf("%d",&qn);
    qdata q[qn];
    for(int i = 0 ; i < qn ; i++) {
        scanf("%d%d",&q[i].i ,&q[i].j);
        q[i].indx = i;
    }

    int result[qn];
    sort(q , q+qn , c1);
    int pos = 0;

    for(int i = 0 ; i<qn ; i++) {
        while (pos < n && pos < q[i].j)
        {
            if (occ[d[pos]] != -1)
            {
                update(occ[d[pos]] + 1, -1);
            }
            update(pos + 1, 1);
            occ[d[pos]] = pos;
            pos++;
        }

        result[q[i].indx] = query(q[i].j) - query(q[i].i - 1);
    }

    for(int i = 0 ; i < qn ; i++)
        printf("%d\n",result[i]);

    return 0;
}